iroh-relay/src/quic.rs

```diff
@@ -0,0 +1,325 @@
+        ///
+        /// # Errors
+        /// If the given `quic_config` contains a [`rustls::ServerConfig`] that cannot
+        /// be converted to a [`QuicServerConfig`], usually because it does not support
+        /// TLS 1.3, this method will error.
+        ///
+        /// # Panics
+        /// If there is a panic during a connection, it will be propagated
+        /// up here. Any other errors in a connection will be logged as a
+        ///  warning.
+        pub(crate) fn spawn(mut quic_config: QuicConfig) -> Result<Self> {
+            quic_config.server_config.alpn_protocols =
+                vec![crate::quic::ALPN_QUIC_ADDR_DISC.to_vec()];
+            let server_config = QuicServerConfig::try_from(quic_config.server_config)?;
+            let mut server_config = quinn::ServerConfig::with_crypto(Arc::new(server_config));
+            let transport_config = Arc::get_mut(&mut server_config.transport).unwrap();
+            transport_config
+                .max_concurrent_uni_streams(0_u8.into())
+                .max_concurrent_bidi_streams(0_u8.into())
+                // enable sending quic address discovery frames
+                .send_observed_address_reports(true);
+
+            let endpoint = quinn::Endpoint::server(server_config, quic_config.bind_addr)?;
+            let bind_addr = endpoint.local_addr()?;
+
+            info!(?bind_addr, "QUIC server bound");
+
+            let cancel = CancellationToken::new();
+            let cancel_accept_loop = cancel.clone();
+
+            let task = tokio::task::spawn(
+                async move {
+                    let mut set = JoinSet::new();
+                    debug!("waiting for connections...");
+                    loop {
+                        tokio::select! {
+                            biased;
+                            _ = cancel_accept_loop.cancelled() => {
+                                break;
+                            }
+                            Some(res) = set.join_next(), if !set.is_empty() => {
+                                if let Err(err) = res {
+                                    // panic if necessary, otherwise, this error has already
+                                    // been logged in `handle_connection`
+                                    if err.is_panic() {
+                                        panic!("task panicked: {:#?}", err);
+                                    }
+                                }
+                            }
+                            res = endpoint.accept() => match res {
+                                Some(conn) => {
+                                     debug!("accepting connection");
+                                     let remote_addr = conn.remote_address();
+                                     set.spawn(async move {
+                                         handle_connection(conn).await
+                                     }.instrument(info_span!("qad-conn", %remote_addr)));
```

Comment:
I think you can write this without wrapping it in another future.  Something like:
```suggestion
                                     set.spawn(
                                         handle_connection(conn)
                                         .instrument(info_span!("qad-conn", %remote_addr))
                                     );
```